File: sdk-java/src/main/java/io/littlehorse/sdk/wfsdk/internal/structdefutil/LHStructProperty.java
package io.littlehorse.sdk.wfsdk.internal.structdefutil;

import io.littlehorse.sdk.common.LHLibUtil;
import io.littlehorse.sdk.common.adapter.LHTypeAdapterRegistry;
import io.littlehorse.sdk.common.exception.LHSerdeException;
import io.littlehorse.sdk.common.proto.StructFieldDef;
import io.littlehorse.sdk.common.proto.TypeDefinition;
import io.littlehorse.sdk.common.proto.VariableValue;
import io.littlehorse.sdk.worker.LHStructField;
import io.littlehorse.sdk.worker.LHStructIgnore;
import java.beans.PropertyDescriptor;
import java.lang.annotation.Annotation;
import java.lang.reflect.Field;
import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;
import java.lang.reflect.RecordComponent;
import java.util.ArrayList;
import java.util.List;
import java.util.Objects;
import java.util.Optional;
import lombok.Getter;
import lombok.extern.slf4j.Slf4j;

@Slf4j
public class LHStructProperty {
    private final PropertyDescriptor pd;
    private final RecordComponent recordComponent;

    @Getter
    private final String propertyName;

    @Getter
    private final String fieldName;

    @Getter
    private final boolean masked;

    @Getter
    private final boolean ignored;

    @Getter
    private final boolean isNullable;

    private final LHStructDefType parentStructDef;

    public LHStructProperty(PropertyDescriptor pd, LHStructDefType parentStructDef) {
        this.pd = Objects.requireNonNull(pd);
        this.recordComponent = null;
        this.parentStructDef = parentStructDef;
        this.propertyName = pd.getName();

        this.fieldName = findFieldName();
        this.masked = findIsMasked();
        this.ignored = findIsIgnored();
        this.isNullable = findIsNullable();
    }

    public LHStructProperty(RecordComponent recordComponent, LHStructDefType parentStructDef) {
        this.pd = null;
        this.recordComponent = Objects.requireNonNull(recordComponent);
        this.parentStructDef = parentStructDef;
        this.propertyName = recordComponent.getName();

        this.fieldName = findFieldName();
        this.masked = findIsMasked();
        this.ignored = findIsIgnored();
        this.isNullable = findIsNullable();
    }

    public VariableValue getValueFrom(Object o) throws LHSerdeException {
        return getValueFrom(o, LHTypeAdapterRegistry.empty());
    }

    public VariableValue getValueFrom(Object o, LHTypeAdapterRegistry typeAdapterRegistry) throws LHSerdeException {
        Method readMethod = getReadMethod();
        if (readMethod == null) {
            throw new IllegalStateException(
                    "No read method for property " + this.fieldName + " found on object of type: " + o.getClass());
        }

        try {
            Object val = readMethod.invoke(o);
            if (val == null) return null;

            if (isNativeArray() && val.getClass().isArray()) {
                return LHLibUtil.objToVarValAsNativeArray(val, getPropertyTypeClass(), typeAdapterRegistry);
            }

            return LHLibUtil.objToVarVal(val, getPropertyTypeClass(), typeAdapterRegistry);
        } catch (LHSerdeException | IllegalAccessException | InvocationTargetException e) {
            throw new LHSerdeException(
                    e, "Failed getting value of property " + this.fieldName + "from object of type: " + o.getClass());
        }
    }

    public void setValueTo(Object o, VariableValue v) throws LHSerdeException {
        setValueTo(o, v, LHTypeAdapterRegistry.empty());
    }

    public void setValueTo(Object o, VariableValue v, LHTypeAdapterRegistry typeAdapterRegistry)
            throws LHSerdeException {
        Method writeMethod = getWriteMethod();
        if (writeMethod == null) {
            throw new IllegalStateException(String.format(
                    "No write method for property [%s] found on object of type [%s]", this.fieldName, o.getClass()));
        }

        try {
            writeMethod.invoke(o, deserializeValue(v, typeAdapterRegistry));
        } catch (IllegalAccessException | InvocationTargetException e) {
            throw new LHSerdeException(
                    e,
                    String.format(
                            "Failed setting value of property [%s] from object of type", this.fieldName, o.getClass()));
        }
    }

    public Object deserializeValue(VariableValue value, LHTypeAdapterRegistry typeAdapterRegistry) {
        return LHLibUtil.varValToObj(value, getPropertyTypeClass(), typeAdapterRegistry);
    }

    /**
     * Converts the LHStructProperty to a StructFieldDef
     * @return a StructFieldDef representing this property
     */
    public StructFieldDef toStructFieldDef() {
        return toStructFieldDef(LHTypeAdapterRegistry.empty());
    }

    public StructFieldDef toStructFieldDef(LHTypeAdapterRegistry typeAdapterRegistry) {
        TypeDefinition typeDef = resolveValidatedFieldType(typeAdapterRegistry);

        StructFieldDef.Builder fieldDef =
                StructFieldDef.newBuilder().setFieldType(typeDef).setIsNullable(isNullable);

        Optional<VariableValue> defaultValue = this.getDefaultValue();
        if (defaultValue.isPresent()) {
            fieldDef.setDefaultValue(defaultValue.get());
        }

        return fieldDef.build();
    }

    private TypeDefinition resolveValidatedFieldType(LHTypeAdapterRegistry typeAdapterRegistry) {
        LHClassType propertyClass;
        try {
            propertyClass = this.getPropertyType(typeAdapterRegistry);

            LHTypeConstraintValidator.ensureNoJsonPrimitiveTypes(propertyClass.getTypeDefinition());
        } catch (IllegalArgumentException | ForbiddenJsonTypeException ex) {
            throw new IllegalArgumentException(
                    String.format(
                            "Invalid StructDef field [%s] on class %s: %s",
                            this.fieldName, this.parentStructDef.getClassType().getCanonicalName(), ex.getMessage()),
                    ex);
        }

        TypeDefinition typeDef = propertyClass.getTypeDefinition().toBuilder()
                .setMasked(this.isMasked())
                .build();

        return typeDef;
    }

    public Optional<VariableValue> getDefaultValue() {
        return getDefaultValue(LHTypeAdapterRegistry.empty());
    }

    public Optional<VariableValue> getDefaultValue(LHTypeAdapterRegistry typeAdapterRegistry) {
        try {
            Object defaultInstance = parentStructDef.createInstance();

            return Optional.ofNullable(getValueFrom(defaultInstance, typeAdapterRegistry));
        } catch (Exception e) {
            log.warn(String.format(
                    "Unable to retrieve default value for Struct Property %s. Blank constructor may not be visible.",
                    this));
            return Optional.empty();
        }
    }

    /**
     * Gets the LHClassType for a given PropertyDescriptor
     * @param pd A PropertyDescriptor representing your Java Beans property
     * @return An optional LHClassType, which will be empty if the property contains the LHStructIgnore annotation on its getter or setters
     */
    public LHClassType getPropertyType() {
        return getPropertyType(LHTypeAdapterRegistry.empty());
    }

    public LHClassType getPropertyType(LHTypeAdapterRegistry typeAdapterRegistry) {
        if (isNativeArray()) {
            return new LHArrayType(getPropertyTypeClass(), typeAdapterRegistry);
        }

        return LHClassType.fromJavaClass(getPropertyTypeClass(), typeAdapterRegistry);
    }

    private boolean isNativeArray() {
        return getPropertyTypeClass().isArray() && !byte[].class.equals(getPropertyTypeClass());
    }

    /// The following methods are used to find annotations on the property, whether they are on the getter, setter, or
    // field itself.
    private <T extends Annotation> T getAnnotation(Class<T> annotationClass) {
        if (this.recordComponent != null && this.recordComponent.isAnnotationPresent(annotationClass)) {
            return this.recordComponent.getAnnotation(annotationClass);
        }

        if ((hasReadMethod() && getReadMethod().isAnnotationPresent(annotationClass))) {
            return getReadMethod().getAnnotation(annotationClass);
        }
        if ((hasWriteMethod() && getWriteMethod().isAnnotationPresent(annotationClass))) {
            return getWriteMethod().getAnnotation(annotationClass);
        }

        return getAnnotationFromField(annotationClass);
    }

    private <T extends Annotation> T getAnnotationFromField(Class<T> annotationClass) {
        for (String fieldName : getCandidateFieldNames()) {
            Field field = getFieldFromClassHierarchy(fieldName);

            if (field != null && field.isAnnotationPresent(annotationClass)) {
                return field.getAnnotation(annotationClass);
            }
        }

        return null;
    }

    private List<String> getCandidateFieldNames() {
        List<String> fieldNames = new ArrayList<>();
        fieldNames.add(propertyName);

        if (getPropertyTypeClass() == boolean.class || getPropertyTypeClass() == Boolean.class) {
            fieldNames.add("is" + Character.toUpperCase(propertyName.charAt(0))
                    + propertyName.substring(1));
        }

        return fieldNames;
    }

    private Field getFieldFromClassHierarchy(String fieldName) {
        Class<?> currentClass = this.parentStructDef.getClassType();

        while (currentClass != null) {
            try {
                return currentClass.getDeclaredField(fieldName);
            } catch (NoSuchFieldException ignored) {
                currentClass = currentClass.getSuperclass();
            }
        }

        return null;
    }

    private boolean findIsIgnored() {
        return getAnnotation(LHStructIgnore.class) != null;
    }

    private boolean findIsMasked() {
        LHStructField lhStructField = getAnnotation(LHStructField.class);

        if (lhStructField == null) return false;

        return lhStructField.masked();
    }

    private String findFieldName() {
        LHStructField lhStructField = getAnnotation(LHStructField.class);

        if (lhStructField == null || lhStructField.name().isBlank()) return propertyName;

        return lhStructField.name();
    }

    private boolean findIsNullable() {
        LHStructField lhStructField = getAnnotation(LHStructField.class);

        if (lhStructField == null) return false;

        return lhStructField.isNullable();
    }

    private boolean hasReadMethod() {
        return getReadMethod() != null;
    }

    private boolean hasWriteMethod() {
        return getWriteMethod() != null;
    }

    private Method getReadMethod() {
        if (this.recordComponent != null) {
            return this.recordComponent.getAccessor();
        }

        return this.pd.getReadMethod();
    }

    private Method getWriteMethod() {
        if (this.recordComponent != null) {
            return null;
        }

        return this.pd.getWriteMethod();
    }

    private Class<?> getPropertyTypeClass() {
        if (this.recordComponent != null) {
            return this.recordComponent.getType();
        }

        return this.pd.getPropertyType();
    }

    @Override
    public String toString() {
        return this.parentStructDef.toString() + ": " + this.fieldName;
    }
}

File: sdk-dotnet/LittleHorse.Sdk/Helper/LHStructProperty.cs
using System.ComponentModel;
using System.Reflection;
using LittleHorse.Sdk.Common.Proto;
using LittleHorse.Sdk.Exceptions;
using LittleHorse.Sdk.Helper;
using LittleHorse.Sdk.Worker;

/// <summary>
/// Represents a single struct property and its LittleHorse metadata.
/// </summary>
public class LHStructProperty
{
  private readonly PropertyDescriptor _pd;
  private readonly LHStructDefType _parentStructDef;
  private readonly PropertyInfo? _propertyInfo;
  private readonly MethodInfo? _getter;
  private readonly MethodInfo? _setter;

  /// <summary>
  /// The LittleHorse field name for this property.
  /// </summary>
  public string FieldName { get; }

  /// <summary>
  /// Whether this field is marked as masked.
  /// </summary>
  public bool Masked { get; }

  /// <summary>
  /// Whether this field should be ignored for struct definitions.
  /// </summary>
  public bool Ignored { get; }

  /// <summary>
  /// Creates a property wrapper for a struct definition.
  /// </summary>
  /// <param name="pd">The property descriptor.</param>
  /// <param name="parentStructDef">The parent struct definition type.</param>
  public LHStructProperty(PropertyDescriptor pd, LHStructDefType parentStructDef)
  {
    ArgumentNullException.ThrowIfNull(pd);
    ArgumentNullException.ThrowIfNull(parentStructDef);
    _pd = pd;
    _parentStructDef = parentStructDef;

    _propertyInfo = _pd.ComponentType.GetProperty(_pd.Name);
    _getter = _propertyInfo?.GetGetMethod(true);
    _setter = _propertyInfo?.GetSetMethod(true);

    FieldName = FindFieldName();
    Masked = FindIsMasked();
    Ignored = FindIsIgnored();
  }

  /// <summary>
  /// Reads the property value from the given object and converts it to a <see cref="VariableValue"/>.
  /// </summary>
  /// <param name="o">The source object.</param>
  /// <returns>The converted variable value, or null if the property value is null.</returns>
  public VariableValue? GetValueFrom(object o)
  {
    if (!HasReadMethod())
    {
      throw new InvalidOperationException(
          "No read method for property " + FieldName + " found on object of type: " + o.GetType());
    }

    try
    {
      object? val = _pd.GetValue(o);
      if (val == null) return null;
      return LHMappingHelper.ObjectToVariableValue(val);
    }
    catch (Exception ex)
    {
      throw new LHSerdeException(
          "Failed getting value of property " + FieldName + " from object of type: " + o.GetType(),
          ex);
    }
  }

  /// <summary>
  /// Sets the property value on the given object using a <see cref="VariableValue"/>.
  /// </summary>
  /// <param name="o">The target object.</param>
  /// <param name="v">The variable value to apply.</param>
  public void SetValueTo(object o, VariableValue v)
  {
    if (!HasWriteMethod())
    {
      throw new InvalidOperationException(string.Format(
          "No write method for property [{0}] found on object of type [{1}]",
          FieldName,
          o.GetType()));
    }

    try
    {
      object? converted = LHMappingHelper.VariableValueToObject(v, _pd.PropertyType);
      _pd.SetValue(o, converted);
    }
    catch (Exception ex)
    {
      throw new LHSerdeException(
          string.Format(
              "Failed setting value of property [{0}] from object of type", FieldName, o.GetType()),
          ex);
    }
  }

  /// <summary>
  /// Converts this property to a <see cref="StructFieldDef"/>.
  /// </summary>
  public StructFieldDef ToStructFieldDef()
  {
    LHClassType propertyClass = GetPropertyType();
    TypeDefinition typeDef = propertyClass.GetTypeDefinition();
    typeDef.Masked = Masked;

    var fieldDef = new StructFieldDef
    {
      FieldType = typeDef
    };

    VariableValue? defaultValue = GetDefaultValue();
    if (defaultValue != null)
    {
      fieldDef.DefaultValue = defaultValue;
    }

    return fieldDef;
  }

  /// <summary>
  /// Gets the default value for this property from a default instance, if available.
  /// </summary>
  public VariableValue? GetDefaultValue()
  {
    try
    {
      object? defaultInstance = _parentStructDef.CreateInstance();
      if (defaultInstance == null)
      {
        return null;
      }

      return GetValueFrom(defaultInstance);
    }
    catch
    {
      return null;
    }
  }

  /// <summary>
  /// Gets the <see cref="LHClassType"/> for this property.
  /// </summary>
  public LHClassType GetPropertyType()
  {
    return LHClassType.FromType(_pd.PropertyType);
  }

  private T? GetAnnotation<T>() where T : Attribute
  {
    if (_propertyInfo != null)
    {
      T? attr = _propertyInfo.GetCustomAttribute<T>(true);
      if (attr != null) return attr;

      if (_getter != null)
      {
        attr = _getter.GetCustomAttribute<T>(true);
        if (attr != null) return attr;
      }

      if (_setter != null)
      {
        attr = _setter.GetCustomAttribute<T>(true);
        if (attr != null) return attr;
      }
    }

    return _pd.Attributes[typeof(T)] as T;
  }

  private bool FindIsIgnored()
  {
    return GetAnnotation<LHStructIgnoreAttribute>() != null;
  }

  private bool FindIsMasked()
  {
    var lhStructField = GetAnnotation<LHStructFieldAttribute>();
    if (lhStructField == null) return false;

    return lhStructField.Masked;
  }

  private string FindFieldName()
  {
    var lhStructField = GetAnnotation<LHStructFieldAttribute>();
    string fieldName;
    if (lhStructField == null || string.IsNullOrWhiteSpace(lhStructField.Name))
    {
      fieldName = _pd.Name;
    }
    else
    {
      fieldName = lhStructField.Name;
    }

    return LowercaseFirstLetter(fieldName);
  }

  private static string LowercaseFirstLetter(string value)
  {
    if (string.IsNullOrEmpty(value)) return value;
    if (char.IsLower(value[0])) return value;

    return char.ToLowerInvariant(value[0]) + value.Substring(1);
  }

  private bool HasReadMethod()
  {
    return _getter != null || !_pd.IsReadOnly;
  }

  private bool HasWriteMethod()
  {
    return _setter != null && !_pd.IsReadOnly;
  }

  /// <summary>
  /// Returns a debug-friendly representation of this property.
  /// </summary>
  public override string ToString()
  {
    return _parentStructDef + ": " + FieldName;
  }
}